pennylane/templates/subroutines/hilbert_schmidt.py

```diff
@@ -0,0 +1,256 @@
+        QuantumFunctionError: The wires ``v_wires`` are a subset of `V` wires.
+        QuantumFunctionError: u_tape and v_tape must act on distinct wires.
+
+    .. UsageDetails::
+
+        Consider that we want to evaluate the Hilbert Schmidt Test cost between the unitary `U` and an approximate
+        unitary `V`. We need to define some functions where it is possible to use the ``.HilbertSchmidt`` template.
+        Here the considered unitary is ``Hadamard`` and we try to compute the cost for the approximate unitary ``RZ``.
+        For an angle which is equal to 0 (identity), we have the maximal cost which is 1.
+
+        .. code-block:: python
+
+            import pennylane as qml
+            from pennylane.templates import HilbertSchmidt
+
+            with qml.tape.QuantumTape(do_queue=False) as U:
+                qml.Hadamard(wires=0)
+
+            def v_circuit(params, v_wires):
+                qml.RZ(params[0], wires=v_wires[0])
+
+            dev = qml.device("default.qubit", wires=4)
+
+            @qml.qnode(dev)
+            def hilbert_test(v_params, v_function, v_wires, u_tape):
+                qml.HilbertSchmidt(v_params, v_function=v_function, v_wires=v_wires, u_tape=u_tape)
+                return qml.probs(u_tape.wires + v_wires)
+
+            def cost_hst(parameters, v_function, v_wires, u_tape):
+                return (1 - hilbert_test(v_params=parameters, v_function=v_function, v_wires=v_wires, u_tape=u_tape)[0])
+
+        Now that you have defined your cost function you can call it for specific parameters.
+
+        >>> cost_hst([0], v_function = v_circuit, v_wires = [1], u_tape = U)
+        1
+
+
+    **Reference**
+
+    [1] Sumeet Khatri, Ryan LaRose, Alexander Poremba, Lukasz Cincio, Andrew T. Sornborger and Patrick J. Coles
+    Quantum-assisted Quantum Compiling.
+    `arxiv/1807.00800 <https://arxiv.org/pdf/1807.00800.pdf>`_
+    """
+    num_wires = AnyWires
+    grad_method = None
+
+    def __init__(self, *params, v_function, v_wires, u_tape, do_queue=True, id=None):
+
+        if not isinstance(u_tape, qml.tape.QuantumTape):
+            raise qml.QuantumFunctionError("The argument u_tape must be a QuantumTape.")
+
+        u_wires = u_tape.wires
+        self.hyperparameters["u_tape"] = u_tape
+
+        if not callable(v_function):
+            raise qml.QuantumFunctionError(
+                "The argument v_function must be a callable quantum function."
+            )
+
+        self.hyperparameters["v_function"] = v_function
+
+        v_tape = qml.transforms.make_tape(v_function)(*params)
+        self.hyperparameters["v_tape"] = v_tape
+        self.hyperparameters["v_wires"] = v_tape.wires
+
+        if len(u_wires) != len(v_wires):
+            raise qml.QuantumFunctionError("U and V must have the same number of wires.")
+
+        if not qml.wires.Wires(v_wires).contains_wires(v_tape.wires):
+            raise qml.QuantumFunctionError("All wires in v_tape must be in v_wires.")
```

Comment:
Just to double-check: `v_tape.wires` is a subset of `v_wires`, but not they are not necessarily equal, correct?

Reply:
Exactly you can imagine that in the function v you start with only a 1 qubit gate but you have two wires.